resources/qtile.1
.\" Man page generated from reStructuredText.
.
.TH "QTILE" "1" "Aug 11, 2020" "0.16.1" "Qtile"
.SH NAME
qtile \- Qtile Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
qtile [\-c config] [\-s qtilesocket] [\-l DEBUG] [\-n]
.SH DESCRIPTION
.sp
\fBqtile\fP runs Qtile on the current $DISPLAY. Qtile is a tiling window manager
written in python. Complete configuration information is available online at
\fI\%http://docs.qtile.org\fP\&.
.SH OPTIONS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.BI \-c \ config\fP,\fB \ \-\-config \ config
Use the specified config file.
.TP
.BI \-s \ qtilesocket\fP,\fB \ \-\-socket \ qtilesocket
Use qtilesocket as the IPC server.
.TP
.BI \-l \ DEBUG\fP,\fB \ \-\-log\-level \ DEBUG
Set the default log level, one of DEBUG, INFO, WARNING, ERROR,
CRITICAL.
.UNINDENT
.UNINDENT
.UNINDENT
.SH FILES
.sp
Qtile searches for configuration files in the following locations:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP 1. 3
The location specified by the \fB\-c\fP option.
.IP 2. 3
\fB$XDG_CONFIG_HOME/qtile/config.py\fP
.IP 3. 3
\fB~/.config/qtile/config.py\fP
.IP 4. 3
The default configuration, distributed as the python module
\fBlibqtile.resources.default_config\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.SH BUGS
.sp
Bugs can be reported to the issue tracker at \fI\%http://github.com/qtile/qtile\fP\&.
.SH AUTHOR
Tycho Andersen
.SH COPYRIGHT
2008-2016, Aldo Cortesi and contributers
.\" Generated by docutils manpage writer.
.


CHANGELOG
qtile 0.16.1, released 2020-08-11:
    !!! Config breakage !!!
	- Hooks 'addgroup', 'delgroup' and 'screen_change' will no longer
	  receive the qtile object as an argument. It can be accessed directly
	  at libqtile.qtile.
    !!! deprecation warning !!!
	- defining a main function in your config is deprecated. You should use
	  @hook.subscribe.startup_complete instead. If you need access to the
	  qtile object, import it from libqtile directly.
    * bugfixes
        - include tests in the release for distros to consume
        - don't resize 0th screen incorrectly on root ConfigureNotify
        - expose qtile object as libqtile.qtile (note that we still consider
          anything not prefixed with cmd_ to be a private API)
        - fix transparent borders
        - MonadTall, MonadWide, and TreeTab now work with Slice

qtile 0.16.0, released 2020-07-20:
    !!! Config breakage !!!
        - Imports from libqtile.widget are now made through a function
          proxy to avoid the side effects of importing all widgets at
          once. If you subclass a widget in your config, import it from
          its own module.
          e.g. from libqtile.widget.pomodoro import Pomodoro
    * features
        - added `guess_terminal` in utils
        - added keybinding cheet sheet image generator
        - custom keyboardlayout display
        - added native support for key chords
        - validate config before restart and refuse to restart with a bad
          config
        - added a bunch of type annotations to config objects (more to come)
    * bugfixes
        - major focus rework; Java-based IDEs such as PyCharm, NetBrains, etc.
          now focus correctly
        - fix a bug where spotify (or any window with focus-to=parent) was
          closed, nothing would be focused and no hotkeys would work
        - support windows unsetting the input hint
        - respects window's/user's location setting if present (WM_SIZE_HINTS)
        - fixed YahooWeather widget for new API
        - fix a bug where _NET_WM_DESKTOPS wasn't correctly updated when
          switching screens in some cases
        - fix a crash in the BSP layout
        - fix a stacktrace when unknown keysyms are encounted
        - make qtile --version output more sane
        - fix a rendering issue with special characters in window names
        - keyboard widget no longer re-sets the keyboard settings every second
        - fix qtile-top with the new IPC model
        - Image widget respects its background setting now
        - correctly re-draw non-focused screens on qtile restart
        - fix a crash when decoding images
        - fix the .when() constraint for lazy objects

qtile 0.15.1, released 2020-04-14
    * bugfixes
	- fix qtile reload (it was crashing)

qtile 0.15.0, released 2020-04-12:
    !!! Config breakage !!!
        - removed the mpd widget, which depended on python-mpd.
	- the Clock widget now requires pytz to handle timezones that are
	  passed as string
	- libqtile.command.Client does not exist anymore and has been
	  replaced by libqtile.command_client.CommandClient
    !!! deprecation warning !!!
	- libqtile.command.lazy is deprecated in favor of libqtile.lazy.lazy
    * features
	- Python 3.8 support
	- `wallpaper` and `wallpaper_mode` for screens
	- bars can now have margins
	- `lazy.toscreen` called twice will now toggle the groups
	  (optional with the `toggle` parameter)
	- `lazy.window.togroup` now has `switch_group` parameter to follow
	  the window to the group it is sent to
	- qtile now copies the default config if the config file does not exist
	- all widgets now use Pango markup by default
	- add an `fmt` option for all textbox widgets
	- new PulseVolume widget for controlling PulseAudio
	- new QuickExit widget, mainly for the default config
	- new non-graph CPU widget
	- KeyboardLayout widget: new `options` parameter
	- CheckUpdates widget: support ArchLinux yay
	- GroupBox widget: new `block_highlight_text_color` parameter
	- Mpd2 widget: new `color_progress` parameter
	- Maildir widget can now display the inbox grand total
	- the Net widget can now use bits as unit
	- Spacer widget: new `background_color` parameter
	- More consistent resize behavior in Columns layout
	- various improvements of the default config
	- large documentation update and improvements (e.g. widget
	  dependencies)
    * bugfixes
	- qtile binary: don't fail if we can't set the locale
	- don't print help if qtile-cmd function returns nothing
	- Monad layout: fix margins when flipped

qtile 0.14.2, released 2019-06-19:
    * bugfixes
        - previous release still exhibited same issues with package data,
          really fix it this time

qtile 0.14.1, released 2019-06-19:
    * bugfixes
        - properly include png files in the package data to install included
          icons

qtile 0.14.0, released 2019-06-19:
    !!! Python version breakage !!!
        - Python 2 is no longer supported
        - Python 3.4 and older is no longer supported
    !!! Config breakage !!!
        - Many internal things were renamed from camel case to snake case. If
          your config uses main(), or any lazy.function() invocations that
          interact directly with the qtile object, you may need to forward port
          them. Also note that we do *not* consider the qtile object to be a
          stable api, so you will need to continue forward porting these things
          for future refactorings (for wayland, etc.). A better approach may be
          to add an upstream API for what you want to do ;)
        - Maildir's subFolder and maildirPath changed to maildir_path and
          sub_folder.
        - the graph widget requires the psutil library to be installed
    * features
        - add custom `change_command` to backlight widget
        - add CommandSet extension to list available commands
        - simplify battery monitoring widget interface and add freebsd
          compatible battery widget implementation

libqtile/scripts/qtile.py
# Copyright (c) 2008, Aldo Cortesi. All rights reserved.
# Copyright (c) 2011, Florian Mounier
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

# Set the locale before any widgets or anything are imported, so any widget
# whose defaults depend on a reasonable locale sees something reasonable.
import locale
import logging
from os import getenv, makedirs, path

from libqtile import confreader
from libqtile.backend.x11 import xcore
from libqtile.log_utils import init_log, logger

try:
    locale.setlocale(locale.LC_ALL, locale.getdefaultlocale())  # type: ignore
except locale.Error:
    pass


try:
    import pkg_resources
    VERSION = pkg_resources.require("qtile")[0].version
except (pkg_resources.DistributionNotFound, ImportError):
    VERSION = 'dev'


def rename_process():
    """
    Try to rename the qtile process if py-setproctitle is installed:

    http://code.google.com/p/py-setproctitle/

    Will fail silently if it's not installed. Setting the title lets you do
    stuff like "killall qtile".
    """
    try:
        import setproctitle
        setproctitle.setproctitle("qtile")
    except ImportError:
        pass


def make_qtile():
    from argparse import ArgumentParser
    parser = ArgumentParser(
        description='A full-featured, pure-Python tiling window manager.',
        prog='qtile',
    )
    parser.add_argument(
        '--version',
        action='version',
        version=VERSION,
    )
    parser.add_argument(
        "-c", "--config",
        action="store",
        default=path.expanduser(path.join(
            getenv('XDG_CONFIG_HOME', '~/.config'), 'qtile', 'config.py')),
        dest="configfile",
        help='Use the specified configuration file',
    )
    parser.add_argument(
        "-s", "--socket",
        action="store",
        default=None,
        dest="socket",
        help='Path of the Qtile IPC socket.'
    )
    parser.add_argument(
        "-n", "--no-spawn",
        action="store_true",
        default=False,
        dest="no_spawn",
        help='Avoid spawning apps. (Used for restart)'
    )
    parser.add_argument(
        '-l', '--log-level',
        default='WARNING',
        dest='log_level',
        choices=('DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'),
        help='Set qtile log level'
    )
    parser.add_argument(
        '--with-state',
        default=None,
        dest='state',
        help='Pickled QtileState object (typically used only internally)',
    )
    options = parser.parse_args()
    log_level = getattr(logging, options.log_level)
    init_log(log_level=log_level)
    kore = xcore.XCore()

    if not path.isfile(options.configfile):
        try:
            makedirs(path.dirname(options.configfile), exist_ok=True)
            from shutil import copyfile
            default_config_path = path.join(path.dirname(__file__),
                                            "..",
                                            "resources",
                                            "default_config.py")
            copyfile(default_config_path, options.configfile)
            logger.info('Copied default_config.py to %s', options.configfile)
        except Exception as e:
            logger.exception('Failed to copy default_config.py to %s: (%s)',
                             options.configfile, e)

    config = confreader.Config(options.configfile, kore=kore)

    # XXX: the import is here because we need to call init_log
    # before start importing stuff
    from libqtile.core import session_manager
    return session_manager.SessionManager(
        kore,
        config,
        fname=options.socket,
        no_spawn=options.no_spawn,
        state=options.state,
    )


def main():
    rename_process()
    q = make_qtile()
    try:
        q.loop()
    except Exception:
        logger.exception('Qtile crashed')
    logger.info('Exiting...')
